### PR TITLE
feat: add C++ SDK

### PR DESCRIPTION
**Pull request by Json-L**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

### CODE REVIEW

**Review by vagetablechicken — APPROVED** [review]

File: src/sdk/openmldb_api.h [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+};
+
+// OpenmldbHandler is used to link openmldb server. to use openmldb, you must create an object of this type
+// cluster :
+// constructor : first parameter         string      format : "IP:Port"    eg : "127.0.0.1:2181"
+//                    second parameter    string      format : "path"       eg : "/openmldb"
+// standalone :
+// constructor : first parameter         string      format : "IP"           eg : "127.0.0.1"
+//                    second parameter    string      format : "Port"        eg : 6527
+class OpenmldbHandler {
```
Comment (vagetablechicken) [inline src/sdk/openmldb_api.h:98]: no namespace?

File: src/sdk/openmldb_api.h [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    Date(int32_t _year, int32_t _month, int32_t _day) : year(_year), month(_month), day(_day) {}
+    ~Date() {}
+    int32_t get_year() const;
+    int32_t get_month() const;
+    int32_t get_day() const;
+
+ private:
+    int32_t year;
+    int32_t month;
+    int32_t day;
```
Comment (dl239) [inline src/sdk/openmldb_api.h]: the name of class members should end with `_`

File: src/sdk/openmldb_api.cc [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    return (handler.get_status())->IsOK();
+}
+
+bool execute(const OpenmldbHandler& handler, const std::string& db, const std::string& sql) {
+    auto rs = (handler.get_router())->ExecuteSQL(db, sql, handler.get_status());
+    if (rs != NULL) resultset_last = rs;
+    return (handler.get_status())->IsOK();
+}
+
+bool execute_parameterized(const OpenmldbHandler& handler, const std::string& db, const std::string& sql,
```
Comment (dl239) [inline src/sdk/openmldb_api.cc:317]: why not set the `execute_parameterized` and `execute` as the function members of `OpenmldbHandler`

File: src/sdk/openmldb_api.h [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+//                    second parameter      string                        name of database
+//                    third parameter          string                        SQL statement
+// insert data into objects of type requets_row using operator '<<'     eg : req << 1 << "hello";
+// reset() is used to clear the inserted data in an object of type RequestRow
+class RequestRow {
+ public:
+    RequestRow(OpenmldbHandler* _handler, const std::string& _db, const std::string& _sql);
+    std::shared_ptr<openmldb::sdk::SQLRequestRow> get_request_row() const;
+    OpenmldbHandler* get_handler() const;
+    std::string get_db() const;
```
Comment (dl239) [inline src/sdk/openmldb_api.h]: return const reference

File: src/sdk/openmldb_api.cc [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    return *this;
+}
+
+ParameterRow& ParameterRow::operator<<(const double value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeDouble);
+    record.push_back(value);
+    return *this;
+}
+
+ParameterRow& ParameterRow::operator<<(const std::string value) {
```
Comment (dl239) [inline src/sdk/openmldb_api.cc]: std::string&

File: src/sdk/openmldb_api.cc [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    auto rs = (handler.get_router())->ExecuteSQLParameterized(db, sql, pr, handler.get_status());
+    if (rs != NULL) resultset_last = rs;
+    return (handler.get_status())->IsOK();
+}
+
+bool execute_request(const RequestRow& req) {
+    auto rr = req.get_request_row();
+    auto rs = ((req.get_handler())->get_router())
+                  ->ExecuteSQLRequest(req.get_db(), req.get_sql(), rr, (req.get_handler())->get_status());
+    if (rs != NULL) resultset_last = rs;
```
Comment (dl239) [inline src/sdk/openmldb_api.cc:329]: guarantee thread safe

File: src/sdk/openmldb_api.h [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+// the symbol "?" to express
+
+// ParameterRow is used to create a parameter line. insert data into objects of type ParameterRow, and replace the "?"
+// constructor : only one parameter of type OpenmldbHandler*
+// insert data into objects of type ParameterRow using operator '<<'     eg : para << 1 << "hello";
+// reset() is used to clear the inserted data in an object of type ParameterRow
+class ParameterRow {
+ public:
+    explicit ParameterRow(const OpenmldbHandler* _handler);
+    std::shared_ptr<openmldb::sdk::SQLRequestRow> get_parameter_row() const;
```
Comment (dl239) [inline src/sdk/openmldb_api.h:127]: do not expose the function that user will not use

**Review by vagetablechicken — CHANGES_REQUESTED** [review]
failed at link sdk lib to a test bin, needs  fix

File: src/sdk/openmldb_api.cc (hunk 3 above)
Comment (vagetablechicken) [reply src/sdk/openmldb_api.cc:317]: copy the style of mysql c++ api and other libs

File: src/sdk/openmldb_api.h (hunk 7 above)
Comment (vagetablechicken) [reply src/sdk/openmldb_api.h:127]: But execute_parameterized need to use it, so keep it

File: src/sdk/openmldb_api.cc (hunk 6 above)
Comment (vagetablechicken) [reply src/sdk/openmldb_api.cc:329]: This style `resultset_last`, can't do thread safe, if 2 threads both set it, thread 1 may can't get the right rs, cuz thread 2 overwrite it.

Comment (vagetablechicken) [reply src/sdk/openmldb_api.cc:329]: If we need it, let's fix it later

**Review by vagetablechicken — APPROVED** [review]

**Comment by vagetablechicken** [comment]
> failed at link sdk lib to a test bin, needs fix

checked.

File: src/sdk/openmldb_api.h (hunk 1 above)
Comment (vagetablechicken) [reply src/sdk/openmldb_api.h:98]: later

**Review by dl239 — APPROVED** [review]

**Pull request merged**